### PR TITLE
Control of datagramWritable Queue

### PR DESCRIPTION
**Pull request by aboba**
Fixes https://github.com/w3c/webtransport/issues/152

***
<a href="https://pr-preview.s3.amazonaws.com/aboba/webtransport/pull/179.html" title="Last updated on Feb 3, 2021, 1:39 AM UTC (fea30fb)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webtransport/179/b8d1a94...aboba:fea30fb.html" title="Last updated on Feb 3, 2021, 1:39 AM UTC (fea30fb)">Diff</a>

### CODE REVIEW

**Comment by yutakahirano** [comment]
@ricea 

Thank you for working on this!

From our implementation experience we cannot use [the high water mark concept in the streams spec](https://streams.spec.whatwg.org/#high-water-mark) because if we use that we would have a queue in the WritableStream and  the underlying sink wouldn't know when the chunk is enqueued, and wouldn't be able to calculate the expiration correctly.

**Comment by aboba** [comment]
@yutakahirano The reference to the high water mark concept has been removed.  Is it accurate now?

**Comment by jan-ivar** [comment]
> From our implementation experience we cannot use [the high water mark concept in the streams spec](https://streams.spec.whatwg.org/#high-water-mark) because if we use that we would have a queue in the WritableStream and the underlying sink wouldn't know when the chunk is enqueued, and wouldn't be able to calculate the expiration correctly.

Can we put queues in both the WritableStream and the underlying source?

An underlying source can then fill its underlying queue by [resolving sinkWritePromises](https://streams.spec.whatwg.org/#writable-stream-default-controller-process-write) prematurely: _"Upon fulfillment of sinkWritePromise, ... Perform ! [DequeueValue](https://streams.spec.whatwg.org/#dequeue-value)(controller)."_ I.e. resolve promises immediately until the underlying buffer fills up, then start returning pending promises, to begin filling up the WritableStream's upper buffer instead, which causes upstream backpressure thanks to `highWaterMark`. Then over time resolve those pending promises one by one as the underlying buffer is emptied to keep it filled.

The size of the underlying queue would be the minimal needed for throughput to not suffer. The size of upper WritableStream queue I don't think matters much as long as it's > 0.

File: index.bs [hunk 1]
```diff
@@ -467,6 +468,11 @@ To <dfn>initialize WebTransport over HTTP</dfn> for a given |transport| and
 
 ## Methods ##  {#webtransport-methods}
 
+: <dfn for="WebTransport" method>setDatagramWritableQueueExpirationDuration()</dfn>
+:: Sets the expiration of the {{datagramWritable}} queue, in milliseconds.
+   {{writer}}.{{ready}} is not fulfilled when there are more outgoing datagrams
+   neither sent nor discarded due to a timeout.
```
Comment (jan-ivar) [inline index.bs:474]: The `ready` API is defined in algorithms in the streams spec, so we need to influence inputs to those algorithms. https://github.com/w3c/webtransport/issues/185

**Review by jan-ivar — CHANGES_REQUESTED** [review]

**Comment by jan-ivar** [comment]
With back pressure, I don't think we'll need `setDatagramWritableQueueExpirationDuration`, since from https://github.com/w3c/webtransport/issues/187#issuecomment-773010072 it's the receiver we'll need to drop packets on, if anything. So I would vote against merging this.

**Comment by yutakahirano** [comment]
@xujiasci, you opened #152. Do you have thoughts on https://github.com/w3c/webtransport/pull/179#issuecomment-773284028?

**Comment by aboba** [comment]
Let's not confuse the behavior of the  QUIC datagram sender implementation with what is going on in Javascript. As I understand it, `setDatagramWritableQueueExpirationDuration()` is about the behavior of the QUIC datagram sender.  The goal was to implement the behavior described in the [QUIC datagram specification Section 5.3](https://tools.ietf.org/html/draft-ietf-quic-datagram-01#section-5.3): 

   `DATAGRAM frames employ the QUIC connection's congestion controller.
   As a result, a connection may be unable to send a DATAGRAM frame
   generated by the application until the congestion controller allows
   it [I-D.ietf-quic-recovery].  The sender implementation MUST
   either delay sending the frame until the controller allows it or drop
   the frame without sending it (at which point it MAY notify the
   application).

   Implementations can optionally support allowing the application to
   specify a sending expiration time, beyond which a congestion-
   controlled DATAGRAM frame ought to be dropped without transmission.`

**Comment by jan-ivar** [comment]
> The sender implementation MUST either delay sending the frame until the controller allows it

I think we should do that. I think this is a problem only with poorly implemented producers that don't listen to back pressure.

I think it's on the receiving end that we need to drop packets, not here.

**Comment by aboba** [comment]
Delaying frames until the controller allows it could cause the frames to lose timeliness, so that they are no longer worth sending. For example, consider an audio or video payload that becomes useless if delayed too long.

File: index.bs [hunk 2]
```diff
@@ -363,6 +363,7 @@ interface WebTransport {
   readonly attribute WebTransportState state;
   readonly attribute Promise&lt;undefined&gt; ready;
   readonly attribute Promise&lt;WebTransportCloseInfo&gt; closed;
+  undefined setDatagramWritableQueueExpirationDuration(long duration);
```
Comment (jan-ivar) [inline index.bs:366]: @yutakahirano suggested moving this to `datagrams`...

Maybe `datagrams.setSendQueueExpiration(long duration)`?

**Comment by jan-ivar** [comment]
Meeting:
- Even with backpressure, stale data may end up in the send queue, due to congestion control/ACKs
- Flushing the send queue of stale data is still useful to say "don't send data more stale than x seconds"

File: index.bs (hunk 2 above)
Comment (aboba) [reply index.bs:366]: That makes sense. Should we close this PR and add the method somewhere else?

**Pull request closed without merging**